persist/sqlite/sectors.go

```diff
@@ -173,234 +180,148 @@ func (s *Store) SectorReferences(root types.Hash256) (refs storage.SectorReferen
+	rows, err := tx.Query(query, encode(lastAccess), sqlSectorBatchSize)
 	if err != nil {
-		return fmt.Errorf("failed to update volume: %w", err)
-	} else if used < 0 {
-		panic("volume usage is negative") // developer error
-	} else if err = incrementNumericStat(tx, metricPhysicalSectors, delta, time.Now()); err != nil {
-		return fmt.Errorf("failed to update metric: %w", err)
+		return nil, fmt.Errorf("failed to select volume sectors: %w", err)
 	}
-	return nil
-}
-
-func pruneSectors(tx *txn, ids []int64) (pruned []types.Hash256, err error) {
-	hasContractRefStmt, err := tx.Prepare(`SELECT EXISTS(SELECT 1 FROM contract_sector_roots WHERE sector_id=$1)`)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare contract reference query: %w", err)
-	}
-	defer hasContractRefStmt.Close()
-
-	hasV2ContractRefStmt, err := tx.Prepare(`SELECT EXISTS(SELECT 1 FROM contract_v2_sector_roots WHERE sector_id=$1)`)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare v2 contract reference query: %w", err)
-	}
-	defer hasV2ContractRefStmt.Close()
-
-	hasTempRefStmt, err := tx.Prepare(`SELECT EXISTS(SELECT 1 FROM temp_storage_sector_roots WHERE sector_id=$1)`)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare temp reference query: %w", err)
-	}
-	defer hasTempRefStmt.Close()
-
-	hasLockStmt, err := tx.Prepare(`SELECT EXISTS(SELECT 1 FROM locked_sectors WHERE sector_id=$1)`)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare lock reference query: %w", err)
-	}
-	defer hasLockStmt.Close()
-
-	clearVolumeStmt, err := tx.Prepare(`UPDATE volume_sectors SET sector_id=NULL WHERE sector_id=$1 RETURNING volume_id`)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare volume reference query: %w", err)
-	}
-	defer clearVolumeStmt.Close()
+	defer rows.Close()
 
-	deleteSectorStmt, err := tx.Prepare(`DELETE FROM stored_sectors WHERE id=$1 RETURNING sector_root`)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare delete sector query: %w", err)
+	for rows.Next() {
+		var ref volumeSectorRef
+		if err := rows.Scan(&ref.VolumeSectorID, &ref.VolumeID); err != nil {
+			return nil, fmt.Errorf("failed to scan volume sector: %w", err)
+		}
+		refs = append(refs, ref)
 	}
-	defer deleteSectorStmt.Close()
+	return
+}
 
-	volumeDelta := make(map[int64]int)
-	for _, id := range ids {
-		var exists bool
-		err := hasContractRefStmt.QueryRow(id).Scan(&exists)
-		if err != nil {
-			return nil, fmt.Errorf("failed to check contract references: %w", err)
-		} else if exists {
-			continue // sector has a contract reference
+// PruneSectors removes volume references for sectors that have not been accessed since the provided
+// timestamp and are no longer referenced by a contract or temp storage.
+func (s *Store) PruneSectors(ctx context.Context, lastAccess time.Time) error {
+	// note: last access can be removed after v2 when sectors are immediately committed to temp storage
+	for i := 0; ; i++ {
```

Comment:
`nit` `for { ... }` debug leftover I'm guessing

Reply:
Consider a 10TB contract expiring, that's 2.6 million records that need to be deleted. The prune operation is batched into small chunks to prevent a single transaction locking the database for a long period. The `for` loop should continue looping until the transaction returns `sql.ErrNoRows`